File: MarkovText/src/StringBasedMarkovTextGenerator.cs
using System.Text;
using System.Text.RegularExpressions;

namespace MarkovText;

/// <summary>
/// Class that generates text based on the Markov chain algorithm
/// </summary>
public partial class StringBasedMarkovTextGenerator : IGenerator
{
    // Safety limit for longest sentence that can be generated, to prevent infinite loops
    public int MaxWordCount = 1000;

    // The order of the Markov chain (how many words in the "state" of the chain)
    private int Order;

    // Default file path for the corpus text
    public const string DefaultCorpusPath = "Resources/thecorsetandthecrinoline.txt";

    // Phrases at the start of sentences are the initial states of the Markov chain
    private readonly List<string> SentenceStarterPhrases = new();

    // Maps prefix word phrases to suffix phrases, e.g., "the big dog" => "big dog was"
    // Tuple also holds the last word of the suffix phrase, e.g., "was"
    private readonly Dictionary<string, List<Tuple<string,string>>> PhraseTransitions = new();

    // Thread-local StringBuilder to avoid memory overhead from multiple threads
    private readonly ThreadLocal<StringBuilder> threadLocalStringBuilder = new(() => new StringBuilder());

    // Sentence delimiters used to detect sentence boundaries
    private static readonly char[] SentenceDelimiters = { '.', '?', '!' };

    // Regex patterns for sentence splitting, sanitization, and whitespace normalization
    [GeneratedRegex(@"\[.+?\]|\""|\)|\(|\'|\n|\r|“|”|’|_")]
    private static partial Regex SanitizerRegex();

    [GeneratedRegex("\n")]
    private static partial Regex FixLineEndingsRegex();

    [GeneratedRegex(@"\s+")]
    private static partial Regex MultipleWhitespaceRegex();

    public override string ToString() => "StringBasedMarkovTextGenerator";

    public void BuildMarkovModel(string corpus, int order = 2)
    {
        Order = order;

        SentenceStarterPhrases.Clear();
        PhraseTransitions.Clear();

        AnalyzeCorpus(corpus);  // Analyze the corpus and build the Markov model

        if (SentenceStarterPhrases.Count == 0)
        {
            throw new ArgumentException($"No phrases of order {Order} could be generated from the corpus: {corpus}");
        }
    }

    public string GenerateSentence(IRandomNumberGenerator random)
    {
        if (SentenceStarterPhrases.Count == 0)
        {
            throw new InvalidOperationException($"There is no Markov model. You need to call {nameof(BuildMarkovModel)} first.");
        }

        var stringBuilder = threadLocalStringBuilder.Value;
        stringBuilder!.Clear();  // Clear the StringBuilder for reuse
        var wordCount = Order;  // Track the current word count to prevent infinite loops
        var phrase = SentenceStarterPhrases.Random(random); // Choose a random starter key from the available starter keys
        stringBuilder.Append(phrase);  // Write the entire sentence starter phrase

        // Continuously generate words based on the Markov chain
        while (PhraseTransitions.TryGetValue(phrase, out var possibleTransitions))
        {
            if (++wordCount >= MaxWordCount)    // Safety check to prevent infinite loops
            {
                throw new SentenceOverflowException($"Word limit {wordCount} reached for sentence:\n{stringBuilder}");
            }

            (phrase, var lastWordInPhrase) = possibleTransitions.Random(random);

            stringBuilder.Append(' ');
            stringBuilder.Append(lastWordInPhrase);   // Write the last word of the phrase to the generated text
        }

        return stringBuilder.ToString();  // Return the generated Markov text
    }

    private void AnalyzeCorpus(string corpus)
    {
        // Replace line endings (e.g., poem line breaks) with a space
        corpus = FixLineEndingsRegex().Replace(corpus, " ");

        // Remove unwanted characters like page numbers, quotes, parentheses, etc.
        corpus = SanitizerRegex().Replace(corpus, "");

        // Normalize multiple consecutive spaces into a single space
        corpus = MultipleWhitespaceRegex().Replace(corpus, " ");

        var wordCount = 0;
        var slidingWindow = new CyclicArray<string>(Order);
        string? previousPhraseString = null;

        foreach (var word in corpus.Trim().Split(' '))  // Split the corpus into words
        {
            if (string.IsNullOrWhiteSpace(word))
            {
                continue;
            }

            slidingWindow[wordCount] = word;

            if (++wordCount < Order)
            {
                if (SentenceDelimiters.Contains(word[^1]))
                {
                    previousPhraseString = null;
                    wordCount = 0;
                }

                continue;
            }

            var phrase = slidingWindow.CreateOffsetArray(wordCount);
            var phraseString = string.Join(' ', phrase);

            if (previousPhraseString == null)
            {
                SentenceStarterPhrases.Add(phraseString);
            }
            else
            {
                PhraseTransitions.AddToList(previousPhraseString, new Tuple<string, string>(phraseString, word));
            }

            previousPhraseString = phraseString;

            if (SentenceDelimiters.Contains(word[^1]))
            {
                previousPhraseString = null;
                wordCount = 0;
            }
        }
    }
}

File: RustImplementation/src/string_based_markov_text_generator.rs
pub mod markov {
    use crate::cyclic_array::markov::CyclicArray;
    use crate::random_number_generator::RandomNumberGenerator;
    use anyhow::anyhow;
    use regex::Regex;
    use std::collections::HashMap;

    const MAX_WORD_COUNT: usize = 1000;

    pub struct StringBasedMarkovTextGenerator {
        order: usize,
        sentences_starter_phrases: Vec<String>,
        phrase_transitions: HashMap<String, Vec<(String, String)>>,
        sentence_delimiters: Vec<char>,
    }

    impl StringBasedMarkovTextGenerator {
        pub(crate) fn generate_sentence(
            &self,
            mut random: impl RandomNumberGenerator,
        ) -> Result<String, anyhow::Error> {
            if self.sentences_starter_phrases.is_empty() {
                return Err(anyhow!(
                    "There is no markov model, call build_markov_model first"
                ));
            }

            let mut string = String::new();
            let mut word_count = self.order;
            let mut phrase = self.sentences_starter_phrases
                [random.next_u32() as usize % self.sentences_starter_phrases.len()]
            .clone();
            string.push_str(&phrase);

            while let Some(phrase_transition) = self.phrase_transitions.get(&phrase) {
                word_count += 1;
                if word_count >= MAX_WORD_COUNT {
                    return Err(anyhow!(
                        "Word limit reached {word_count} reached for sentence: {string}"
                    ));
                }

                let randomu32 = random.next_u32();
                let len = phrase_transition.len();
                let index = randomu32 as usize % len;

                let (s, last_word_in_phrase) = phrase_transition.get(index).unwrap();

                phrase = s.clone();

                string.push(' ');
                string.push_str(last_word_in_phrase);
            }

            Ok(string)
        }
    }

    impl StringBasedMarkovTextGenerator {

        pub fn new() -> StringBasedMarkovTextGenerator {
            StringBasedMarkovTextGenerator {
                phrase_transitions: HashMap::new(),
                order: 0,
                sentences_starter_phrases: vec![],
                sentence_delimiters: vec![',', '.', '!'],
            }
        }

        pub fn build_markov_model(
            &mut self,
            corpus: &str,
            order: usize,
        ) -> Result<(), anyhow::Error> {
            self.order = order;

            self.analyze_corpus(corpus)?;

            if self.phrase_transitions.is_empty() {
                return Err(anyhow!("No phrases found in the corpus"));
            }

            if self.sentences_starter_phrases.is_empty() {
                return Err(anyhow!(
                    r"No phrases of order {} could be generated from the corpus: {}",
                    self.order,
                    corpus
                ));
            }

            Ok(())
        }

        pub fn analyze_corpus(&mut self, corpus: &str) -> Result<(), anyhow::Error> {
            let s = corpus.replace("\n", " ");
            let santize = Regex::new(r#"\[.+?\]|\""|\)|\(|\'|\n|\r|“|”|’|_"#)?;
            let s2 = santize.replace_all(&s, "");
            let cleaned_corpus = Regex::new(r#"\s+"#)?.replace_all(&s2, " ");

            let mut word_count: usize = 0;
            let mut sliding_widow = CyclicArray::new(self.order);
            let mut previous_phrase_string: Option<String> = None;

            for word in cleaned_corpus.split_whitespace() {
                if word.is_empty() {
                    continue;
                }
                sliding_widow[word_count] = word.to_string();

                word_count += 1;
                if word_count < self.order {
                    if word.chars().nth_back(0).is_none() {
                        return Err(anyhow!("Word is empty"));
                    }

                    if self
                        .sentence_delimiters
                        .contains(&word.chars().nth_back(0).unwrap())
                    {
                        previous_phrase_string = None;
                        word_count = 0;
                    }
                    continue;
                }

                let phrase = sliding_widow.create_offset_array(word_count);
                let phrase_string = phrase.join(" ");

                if previous_phrase_string.is_none() {
                    self.sentences_starter_phrases.push(phrase_string.clone());
                } else {
                    self.phrase_transitions
                        .entry(previous_phrase_string.clone().unwrap())
                        .and_modify(|e| e.push((phrase_string.clone(), String::from(word))))
                        .or_insert_with(|| vec![(phrase_string.clone(), String::from(word))]);
                }

                previous_phrase_string = Some(phrase_string);

                if self
                    .sentence_delimiters
                    .contains(&word.chars().last().unwrap())
                {
                    previous_phrase_string = None;
                    word_count = 0;
                }
            }

            Ok(())
        }
    }
}
